File: system-test/src/testkit/driver/port/book-shop-driver.ts
import { Result } from '../../common/result.js';
import { AsyncCloseable } from './async-closeable.js';
import { GoToBookShopRequest } from './dtos/GoToBookShopRequest.js';
import { GoToBookShopResponse } from './dtos/GoToBookShopResponse.js';
import { PlaceOrderRequest } from './dtos/PlaceOrderRequest.js';
import { PlaceOrderResponse } from './dtos/PlaceOrderResponse.js';
import { CancelOrderRequest } from './dtos/CancelOrderRequest.js';
import { CancelOrderResponse } from './dtos/CancelOrderResponse.js';
import { DeliverOrderRequest } from './dtos/DeliverOrderRequest.js';
import { DeliverOrderResponse } from './dtos/DeliverOrderResponse.js';
import { ViewOrderRequest } from './dtos/ViewOrderRequest.js';
import { ViewOrderResponse } from './dtos/ViewOrderResponse.js';
import { SystemError } from './dtos/errors/SystemError.js';
import { PublishCouponRequest } from './dtos/PublishCouponRequest.js';
import { PublishCouponResponse } from './dtos/PublishCouponResponse.js';
import { BrowseCouponsRequest } from './dtos/BrowseCouponsRequest.js';
import { BrowseCouponsResponse } from './dtos/BrowseCouponsResponse.js';

export interface BookShopDriver extends AsyncCloseable {
  goToBookShop(request: GoToBookShopRequest): Promise<Result<GoToBookShopResponse, SystemError>>;
  placeOrder(request: PlaceOrderRequest): Promise<Result<PlaceOrderResponse, SystemError>>;
  cancelOrder(request: CancelOrderRequest): Promise<Result<CancelOrderResponse, SystemError>>;
  deliverOrder(request: DeliverOrderRequest): Promise<Result<DeliverOrderResponse, SystemError>>;
  viewOrder(request: ViewOrderRequest): Promise<Result<ViewOrderResponse, SystemError>>;
  publishCoupon(request: PublishCouponRequest): Promise<Result<PublishCouponResponse, SystemError>>;
  browseCoupons(request: BrowseCouponsRequest): Promise<Result<BrowseCouponsResponse, SystemError>>;
}

File: system-test/src/main/java/com/mycompany/bookshop/testkit/driver/port/BookShopDriver.java
package com.jcupac.bookshop.testkit.driver.port;

import com.jcupac.bookshop.testkit.driver.port.dtos.error.SystemError;
import com.jcupac.bookshop.testkit.driver.port.dtos.BrowseCouponsRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.BrowseCouponsResponse;
import com.jcupac.bookshop.testkit.driver.port.dtos.CancelOrderRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.CancelOrderResponse;
import com.jcupac.bookshop.testkit.driver.port.dtos.DeliverOrderRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.DeliverOrderResponse;
import com.jcupac.bookshop.testkit.driver.port.dtos.GoToBookShopRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.GoToBookShopResponse;
import com.jcupac.bookshop.testkit.driver.port.dtos.PlaceOrderRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.PlaceOrderResponse;
import com.jcupac.bookshop.testkit.driver.port.dtos.PublishCouponRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.PublishCouponResponse;
import com.jcupac.bookshop.testkit.driver.port.dtos.ViewOrderRequest;
import com.jcupac.bookshop.testkit.driver.port.dtos.ViewOrderResponse;
import com.jcupac.bookshop.testkit.common.Result;

public interface BookShopDriver extends AutoCloseable {
    Result<GoToBookShopResponse, SystemError> goToBookShop(GoToBookShopRequest request);

    Result<PlaceOrderResponse, SystemError> placeOrder(PlaceOrderRequest request);

    Result<CancelOrderResponse, SystemError> cancelOrder(CancelOrderRequest request);

    Result<DeliverOrderResponse, SystemError> deliverOrder(DeliverOrderRequest request);

    Result<ViewOrderResponse, SystemError> viewOrder(ViewOrderRequest request);

    Result<PublishCouponResponse, SystemError> publishCoupon(PublishCouponRequest request);

    Result<BrowseCouponsResponse, SystemError> browseCoupons(BrowseCouponsRequest request);
}
